### ISSUE TITLE
Segfault issue in PHP for gcloud/spanner connection

### ISSUE THREAD

**Issue by m-lagarde** [protocolbuffers/protobuf#8366]
**What version of protobuf and what language are you using?**

Version: 3.13+
Language: PHP 7.4+

**What did you do?**

Using `google/cloud-spanner` lib with `grpc` transport
On each connection/access to the database, we get a segmentation fault (`Segmentation fault (core dumped)`)

The last stable version we've found for this usage is the `3.12.4` (tested with each `7.4+` PHP version)

**Comment by haberman** [comment]
Thanks for the report and sorry for the slow reply.

Please try 3.17.1 (released today). A number of memory errors have been fixed since 3.13.

**Comment by m-lagarde** [comment]
Tested with `3.17.3` and still getting the segmentation fault

**Comment by haberman** [comment]
This happens every time you connect to the database?

Can you post more information, like a crash report?

**Comment by haberman** [comment]
@xvzf thanks for the info! Unfortunately this backtrace doesn't make clear how we would have gotten a crash. The crashing function (according to this backtrace) is extremely simple:

```
bool upb_msgdef_mapentry(const upb_msgdef *m) {
  return m->map_entry;
}
```

This could only crash if `m` was an invalid pointer. But the backtrace seems to suggest that `m` is a valid heap address.

Two questions:
1. Does this reproduce if you are not using Blackfire?
2. Could you run `p *m` in the debugger to see if the pointer is readable, and what data is there?

A few other things that could help narrow down the problem, if you're able:
1. If you were able to build the extension in non-optimized mode (`-O0`) that would also be useful (it would make the stack trace more reliable).
2. If you could reproduce the crash under Valgrind, that would be even better. Set the environment variable `USE_ZEND_ALLOC=0` before running PHP to avoid a bunch of spurious Valgrind errors.

gRPC trace/debug logs shouldn't be necessary, but thanks for the offer.

**Comment by xvzf** [comment]
Sure; I tried to debug it further without the blackfire extension;

seems there is indeed an invalid pointer
```
Core was generated by `php /var/www/html/bin/console worker:result-extraction {"body":"{\"id\":\"0b902'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  upb_msgdef_mapentry (m=m@entry=0x7f7bbc0b1d30) at /tmp/pear/temp/protobuf/php-upb.c:5181
5181	/tmp/pear/temp/protobuf/php-upb.c: No such file or directory.
[Current thread is 1 (LWP 29)]
(gdb) backtrace full
#0  upb_msgdef_mapentry (m=m@entry=0x7f7bbc0b1d30) at /tmp/pear/temp/protobuf/php-upb.c:5181
No locals.
#1  0x00007f7bbd348b98 in Descriptor_FromMessageDef (m=0x7f7bbc0b1d30, val=0x7fff2ff0e420) at /tmp/pear/temp/protobuf/def.c:538
        ce = 0x0
        ret = <optimized out>
#2  Descriptor_FromMessageDef (val=0x7fff2ff0e420, m=0x7f7bbc0b1d30) at /tmp/pear/temp/protobuf/def.c:530
        ce = <optimized out>
        ret = <optimized out>
        __z = <optimized out>
#3  0x00007f7bbd348e42 in Descriptor_FromClassEntry (ce=0x7f7bbbf04040, val=0x7fff2ff0e420) at /tmp/pear/temp/protobuf/def.c:558
No locals.
#4  Descriptor_FromClassEntry (ce=0x7f7bbbf04040, val=0x7fff2ff0e420) at /tmp/pear/temp/protobuf/def.c:556
No locals.
#5  Descriptor_GetFromClassEntry (ce=ce@entry=0x7f7bbbf04040) at /tmp/pear/temp/protobuf/def.c:580
        desc = {value = {lval = 1119419264, dval = 5.5306660163529287e-315, counted = 0x42b8fb80, str = 0x42b8fb80, arr = 0x42b8fb80, obj = 0x42b8fb80, res = 0x42b8fb80, ref = 0x42b8fb80, ast = 0x42b8fb80, zv = 0x42b8fb80, ptr = 0x42b8fb80, ce = 0x42b8fb80, func = 0x42b8fb80, ww = {w1 = 1119419264, w2 = 0}}, u1 = {
            type_info = 1, v = {type = 1 '\001', type_flags = 0 '\000', u = {extra = 0}}}, u2 = {next = 153168683, cache_slot = 153168683, opline_num = 153168683, lineno = 153168683, num_args = 153168683, fe_pos = 153168683, fe_iter_idx = 153168683, access_flags = 153168683, property_guard = 153168683,
            constant_flags = 153168683, extra = 153168683}}
#6  0x00007f7bbd3516ad in zim_Message___construct (execute_data=0x7f7bbfc146c0, return_value=<optimized out>) at /tmp/pear/temp/protobuf/message.c:591
        intern = 0x7f7bbcd80120
        desc = <optimized out>
        ce = 0x7f7bbbf04040
        arena = 0x7f7bbbd87c20
        init_arr = 0x0
#7  0x00005578ba0fc509 in execute_ex ()
No symbol table info available.
#8  0x00005578ba087240 in zend_call_function ()
No symbol table info available.
#9  0x00005578b9fa180f in ?? ()
No symbol table info available.
#10 0x00005578ba0fa07f in execute_ex ()
No symbol table info available.
#11 0x00005578ba087240 in zend_call_function ()
No symbol table info available.
#12 0x00005578ba087665 in zend_call_known_function ()
No symbol table info available.
#13 0x00005578ba116074 in zend_objects_destroy_object ()
No symbol table info available.
#14 0x00005578ba11a6e6 in zend_objects_store_call_destructors ()
No symbol table info available.
#15 0x00005578ba0852c3 in ?? ()
No symbol table info available.
#16 0x00005578ba0948e5 in zend_call_destructors ()
No symbol table info available.
#17 0x00005578ba030785 in php_request_shutdown ()
No symbol table info available.
#18 0x00005578ba121340 in ?? ()
No symbol table info available.
#19 0x00005578b9dc068b in ?? ()
No symbol table info available.
#20 0x00007f7bc06bea03 in ?? () from /lib/ld-musl-x86_64.so.1
No symbol table info available.
#21 0x00007f7bc06be9dc in ?? () from /lib/ld-musl-x86_64.so.1
No symbol table info available.
#22 0x00007fff2ff10890 in ?? ()
No symbol table info available.
#23 0x0000000000000000 in ?? ()
No symbol table info available.
(gdb) p *m
Cannot access memory at address 0x7f7bbc0b1d30
```

The full valgrind is here: https://gist.github.com/xvzf/4074f1e7d3b816ac75077e0fa811cfbb

**Comment by haberman** [comment]
Thanks for the extra info, that's very helpful.

I think this is the same as https://github.com/protocolbuffers/protobuf/issues/8454, it appears you are creating a proto message object during request shutdown, perhaps from a __destruct() method?

I'm not sure if we can reasonably make this case work without leaking memory. Here is one thing you could try: set `protobuf.keep_descriptor_pool_after_request` which will make the extension cache this information across requests. That could help avoid this memory from being freed prior to this message constructor.

We should definitely be giving a better error message here rather than segfaulting.

**Comment by xvzf** [comment]
So; good news is I tracked it down, bad news is I don't know how to proceed from here, for me it looks correctly implemented.

https://github.com/googleapis/google-cloud-php-spanner/blob/12994e83c3e0730146b8c6633acf4f57a4202c90/src/Connection/Grpc.php#L745

When commenting out this call, there is no Segfault (-> you were right with the creating a proto message during request shutdown). Could there be a race-condition? If you can hind me to the code that follows next I can try to trace it down further.

**Comment by haberman** [comment]
Hmm, `protobuf.keep_descriptor_pool_after_request=1` may not be functioning as intended if the destructors for global objects (such as `DescriptorPool_destructor`) are running before the [request shutdown function](https://github.com/protocolbuffers/protobuf/blob/master/php/ext/google/protobuf/protobuf.c#L185-L188). This feature is not well-tested.

We may want to move that shutdown code into the `DescriptorPool_destructor`: we can test whether this DescriptorPool is the generated pool (`PROTOBUF_G(generated_pool )`) and `protobuf.keep_descriptor_pool_after_request=1` is enabled, and avoid calling `upb_symtab_free()` in that case (we should cache it in `PROTOBUF_G(saved_symtab)` instead).

**Comment by tomwalder** [comment]
FWIW, we ran into this `__destruct()` conflict as part of an upgrade process. We have been adding the `grpc` and `protobuf` extensions so our existing software stack.

The segfault manifested for us because we had an event collection class with a `__destruct()` method to publish, which works fine with the REST gateway.

`protobuf.keep_descriptor_pool_after_request=1` did not fix the issue.

We've had to re-work the application to support the extension.
